[HEAVY WIP] Core Tourney Code

Ryll-Ryll opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
This is my work on the event's structure for handling the contestants, teams, rounds, and related structural stuff to make the showrunners' lives easier. Compared to the toolbox tourney, this event represents a big step up in scope, needing to handle a variety of different events, each with their own rules for team sizes, win conditions, plus the need for dynamic team creation/destruction. Since my availability for the next month may be spotty, I'm just diving in and slapping this together with all-new datum types so I can dive into building the structures into what I need rather than cram all these upgrades into the existing tourney event code. This code emphasizes being quick to write, decently modular, and as resistant as possible to someone bricking it during showtime, though as a result, elegance takes a backseat design-wise.

The main idea is that the tournament will be run by 3 different datum types that interact with each other. They are, from big to small

1. Roster: There will be one global roster datum, and this will contain all of the teams and contestants, as well as the logic and helpers showrunners will use to mark who is eliminated when, and who's against who. Will likely be interacted with mainly by the arena console like last time, though in this case all the main info lives on the roster datum instead of the console.
2. Team: These are temporary datums that are created and deleted every round, per the rules for that round's challenge (different events need different team sizes/counts, possibly different numbers of teams competing at a time in each round, possibly different rules on only part of a losing team being knocked out, etc). Their role is to contain references to the contestant datums they represent, and handle putting them in a round/recording the results as easy as possible.
3. Contestant: One of these is created for each contestant in the show, and they last the entire time. When still in the game, these are used to handle team grouping and spawning, and once they lose, they're deactivated and swept aside. 

Throwing this PR up now to help motivate myself, as well as so people can track the progress and point out edge-cases I may miss. There will be a *lot* of iteration and changes over the next few weeks, but the plan is for the prototype to be ready for 6/6, leaving the next few weeks for polishing and tuning. I plan to have a checklist of what I need to do to count this as bare-minimum working by the end of Saturday, to better chart out what needs doing.

Ryll-Ryll committed:
aei

Ryll-Ryll committed:
wrlk

Ryll-Ryll committed:
further work on roster, team, and contestant datums

Ryll-Ryll committed:
explaining thought process more

Ryll-Ryll committed:
working on new players connecting

Ryll-Ryll committed:
further work on contestants

CitrusGender commented:
read all of it, great job shaps. Will likely need to add some stuff for the spawning in and management of the actual event but this is going really good. I'll try to make sure that we get it bug tested after I'm back so we can get it perfect.

Ryll-Ryll committed:
beginning work on setting up matches

Ryll-Ryll committed:
my computer is crashing

Ryll-Ryll committed:
checkpoint, continuing work on the arena UI and making matches

Ryll-Ryll committed:
divying up people into teams and assigning 2 teams to a match

Ryll-Ryll committed:
this match and team code is horribly bloated but i'll cut it down soon

Ryll-Ryll committed:
calling this """prototype complete"""

Ryll-Ryll committed:
basic docs and logs for roster.dm

Ryll-Ryll committed:
Basically docs contestant and team datums

Ryll-Ryll committed:
neatens up arena.dm a bit

Ryll-Ryll committed:
divvy not divy

Ryll-Ryll commented:
Calling this "prototype complete", you can now do almost all of the important contestant/team/match structure things like load a roster, divvy people up into teams randomly for each round with variable team sizes/team counts, eliminate people, mark people for elimination based on losing a round, etc. Here's a few random screenies I posted in eventsbus

8 random empty contestant datums to display the system
[![2021-06-05_05-53-34.png](https://i.imgur.com/3j831Yz.png)](https://i.imgur.com/3j831Yz.png)
Randomly divvying up those 8 people into 4 teams of 2
[![2021-06-05_05-53-34.png](https://cdn.discordapp.com/attachments/604145125832261632/850675723848646656/unknown.png)](https://cdn.discordapp.com/attachments/604145125832261632/850675723848646656/unknown.png)
In the match menu, pitting the first two teams against each other
[![2021-06-05_05-53-34.png](https://cdn.discordapp.com/attachments/604145125832261632/850675812008591370/unknown.png)](https://cdn.discordapp.com/attachments/604145125832261632/850675812008591370/unknown.png)
The first match is resolved and team 2 beat team 1, so team 1 is marked for elimination. Next 2 teams are queued up.
[![2021-06-05_05-53-34.png](https://cdn.discordapp.com/attachments/604145125832261632/850676065990606879/unknown.png)](https://cdn.discordapp.com/attachments/604145125832261632/850676065990606879/unknown.png)
Second match is resolved, team 3 beats team 4, this round of matches is now done.
[![2021-06-05_05-53-34.png](https://cdn.discordapp.com/attachments/604145125832261632/850676146114789386/unknown.png)](https://cdn.discordapp.com/attachments/604145125832261632/850676146114789386/unknown.png)
After clicking on confirm elimination for each losing team, this shows the updated contestants list, with 4 still active and 4 now eliminated
[![2021-06-05_05-53-34.png](https://cdn.discordapp.com/attachments/604145125832261632/850676295238942730/unknown.png)](https://cdn.discordapp.com/attachments/604145125832261632/850676295238942730/unknown.png)
And so on, the next round of matches (really only one match, 2 teams with 2 contestants each facing off) is queued up with the survivors
[![2021-06-05_05-53-34.png](https://cdn.discordapp.com/attachments/604145125832261632/850676441582141460/unknown.png)](https://cdn.discordapp.com/attachments/604145125832261632/850676441582141460/unknown.png)

Still more polishing, testing, and nice-to-haves to work on, but at least most of the bare minimum is there! Also have to implement the older arena loading code, and handle spawning people in when their match is up, as far as bare minimum goes.

CitrusGender commented on code/game/machinery/computer/arena.dm:
```diff
@@ -367,6 +583,7 @@
+	*/
 
 /// Arena spawnpoint
 /obj/machinery/arena_spawn
```
One thing to consider, spawning may be loaded into the map directly so it will be different than last event when we loaded people into static zones that didn't change with arena maps. I'm not sure if the previous code was set up to account for that though.

CitrusGender commented:
Just a thought as well, but it may be in good faith to add a button that applies godmode to all spawned contestants that can be toggled on and off. I can add that if need be, but it's just a thought since I don't want some rounds ending extremely early since one team just murders the other. This may be relevant in like the courtroom challenge.

Let me know if you're too swamped. I can help out with that.

Code should just be as simple as getting all mobs spawned and then

`M.status_flags ^= GODMODE`

CitrusGender commented:
Additionally, just make sure that teams have icons on top of themselves that match what color team they have within the machine. I think that players were able to see this during toolbox tournament. This is important since admins should be able to check which teams won even if they're not too familiar with the people playing. 

Obviously no need for an icon when there's no teams. A spawn all contestants button would be helpful for the first round too, unless the current system would allow for that easily.
![image](https://user-images.githubusercontent.com/31262308/120906502-f1578100-c627-11eb-851a-c69bd54eaf8d.png)

CitrusGender commented:
Lastly, (and I promise this is last!!!!) , some mappers have been including spawn protection in their maps with the blue and red spawn protection traps `/obj/structure/trap/ctf/blue` and `/obj/structure/trap/ctf/red` respectively.

We can change these out to fit out event code, hooking the teams up to these spawn protections will likely go a long way.

Let me know if you need me to help out with this.

![image](https://user-images.githubusercontent.com/31262308/120908088-0555af80-c635-11eb-99bd-a8a6ad2450f5.png)

Ryll-Ryll committed:
beginning implementing arena loading

Ryll-Ryll committed:
frozen/godmode toggles, move contestant to own file, random wound toggle

Ryll-Ryll committed:
!= -> = !

Ryll-Ryll committed:
makes toggle freeze/god on teams work

Ryll-Ryll committed:
actually makes freeze/godmode teams work

Ryll-Ryll commented:
Will likely call it a night before too much longer, but I've added options to toggle godmode/freezing in place on everyone/entire teams/individual contestants (though there's no button for individuals in the menus right now). These can be set in advance, so if you set everyone to freeze, for example, any contestant that is spawned in as a person will be frozen in place until unfrozen. 

This way, you can set it so everyone is frozen and godmoded before they are spawned in, then manually lift the freezing and godmoding by team when you're ready for them to go. I'll post pictures of the UI later.

I've also added a toggle for enabling/disabling random wounds to the arena computer (default to random wounds being disabled), though they can still be applied manually without problems like for the medical triage event.

The main things left to do off the top of my head:

- [x] Move arena loading to new code
- [x] Move antag team huds to new code
- [x] Finish button for spawning people in
- [ ] Neaten up UI lots

hurrah

CitrusGender commented:
Looking at the code, it looks like it's keeping with the two spawners so I'm assuming it will be restricted to two teams. This isn't a problem too much, we do have one map that has three teams but we can change that.

Ryll-Ryll committed:
before i update to new

Ryll-Ryll committed:
Merge branch 'gameshow' into eventcode

Ryll-Ryll committed:
spawning and loading arenas baseline working

Ryll-Ryll committed:
team antag huds now work!

Ryll-Ryll committed:
possibly start removing old arena code

Ryll-Ryll commented:
I can't believe it, but antag huds and arena loading have been successfully moved to the new roster datum and menu, spawning contestants by teams and toggling freeze/godmode on them pre-roundstart is working successfully, and things actually seem to be going well! With any luck, we may actually pull this off!

At last, I'm going to be turning my attention towards testing my code, improving the UI, and adding some small QOL features to make the showerunners' jobs easier (especially regarding the battle royale opener). The following barrage of screenshots show roughly what it looks like to load an arena, split 12 contestants into 6 teams of 2, run 3 team vs team matches with a winner and a loser, and then eliminating the losing teams, leaving 6 more contestants for the next round.

(Though I'm still not really sure what the plan is for how you guys are planning on choosing one person from each losing team to get eliminated rather than the whole team, ez support for doing so will be in soon.)

[![dreamseeker_2021-06-18_02-57-26.png](https://i.imgur.com/8ON2WFSl.jpg)](https://i.imgur.com/8ON2WFS.png)

[![dreamseeker_2021-06-18_02-58-04.png](https://i.imgur.com/wzEFdUwl.jpg)](https://i.imgur.com/wzEFdUw.png)

[![2021-06-18_02-58-14.png](https://i.imgur.com/NIwSBYbl.jpg)](https://i.imgur.com/NIwSBYb.png)

[![2021-06-18_02-58-35.png](https://i.imgur.com/f8hFnUKl.jpg)](https://i.imgur.com/f8hFnUK.png)

[![2021-06-18_02-58-50.png](https://i.imgur.com/IawmnUBl.jpg)](https://i.imgur.com/IawmnUB.png)

[![2021-06-18_02-59-03.png](https://i.imgur.com/HtHUPmPl.jpg)](https://i.imgur.com/HtHUPmP.png)

[![2021-06-18_02-59-46.png](https://i.imgur.com/pwzSETzl.jpg)](https://i.imgur.com/pwzSETz.png)

[![2021-06-18_03-00-05.png](https://i.imgur.com/4oekQN6l.jpg)](https://i.imgur.com/4oekQN6.png)

CitrusGender commented:
map was updated btw, see pinned in eventbus.

Ryll-Ryll committed:
adds some ckeys_at_large funcs, adding specific people to teams

CitrusGender commented:
alright seems like we've got something workable. I'll likely test tomorrow after fixing up maps. A personal belief in priority right now uh:

Battle Royal spawning mechanism (Honestly, all it needs is just to use the machine and spawn them to each one. The mechanism for elimination isn't as much of a problem.) >>> Spawn Protection (From CTF) > Having more than two teams > Adding in Dreath's Voice Lines upon elimination

May be more to add after this.

CitrusGender commented:
Also stretch goal (later part), second round of deletion on map deletion.

Ryll-Ryll committed:
new spawn protection obj, br spawning, attempt hardclear arena

Ryll-Ryll committed:
Merge branch 'gameshow' into eventcode

Ryll-Ryll committed:
royale with cheese, not royal

Ryll-Ryll commented:
I already touched base with you obv, but took care of BR spawning, spawn protection object, and the extra round of deletions on arena switching (though that all still needs to be tested). Will look into the 3 team cooking event next, and I'll hook up the voice lines either tonight or tomorrow

Ryll-Ryll committed:
Continued work on battle royale, add _ to datum filenames

Ryll-Ryll committed:
halfway through adding sounds

Ryll-Ryll committed:
sound

Ryll-Ryll committed:
Merge branch 'gameshow' into eventcode

CitrusGender commented:
PR is good, I think we're still missing a few things to test but I'm going to put this on main so that we can have a consistent branch instead of merging two branches into a third one. Still needs some testing but I'm happy with what we have.

CitrusGender committed:
Merge branch 'gameshow' into eventcode
